### PR TITLE
Fixing TestExecutor stuck scenarios

### PR DESCRIPTION
**Pull request by dsvoronin**
PR on top of #673 with actual fix

### CODE REVIEW

File: subprojects/gradle/instrumentation-tests-test-fixtures/src/main/kotlin/com/avito/instrumentation/report/listener/FakeTestReporter.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.avito.instrumentation.report.listener
+
+import com.avito.runner.service.model.TestCase
+import org.funktionale.tries.Try
+import java.io.File
+
+class FakeTestReporter : TestReporter() {
```
Comment (eugene-krivobokov) [inline subprojects/gradle/instrumentation-tests-test-fixtures/src/main/kotlin/com/avito/instrumentation/report/listener/FakeTestReporter.kt:7]: We use two naming patterns - fake and stub. Is there any difference?

File: subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+        executor.execute(
+            application = null,
+            testApplication = File(""),
+            testsToRun = listOf(),
+            executionParameters = ExecutionParameters.createStubInstance(),
+            output = tempDir
+        )
+
+        // no assertion needed: just stuck in runBlocking forever before fix
```
Comment (eugene-krivobokov) [inline subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt:44]: How about adding `timeout` for the Test?

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/executing/TestExecutor.kt [hunk 3]
```diff
@@ -49,6 +49,8 @@ interface TestExecutor {
         ) {
             withDevices(reservations = reservations(testsToRun)) { devices ->
 
+                logger.info("Got devices channel")
```
Comment (eugene-krivobokov) [inline subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/executing/TestExecutor.kt]:
```suggestion
                logger.debug("Got devices channel")
```

**Review by eugene-krivobokov — APPROVED** [review]

File: subprojects/gradle/instrumentation-tests-test-fixtures/src/main/kotlin/com/avito/instrumentation/report/listener/FakeTestReporter.kt (hunk 1 above)
Comment (sergioboi) [reply subprojects/gradle/instrumentation-tests-test-fixtures/src/main/kotlin/com/avito/instrumentation/report/listener/FakeTestReporter.kt:7]: In that case, it is an EmptyTestReporter

File: subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt (hunk 2 above)
Comment (sergioboi) [reply subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt:44]: Important

File: subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    @Test
+    fun `executor - finished - problem with deployment`(@TempDir tempDir: File) {
```
Comment (sergioboi) [inline subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt]: Could you make the problem explicit I don't understand that problem in the test body

File: subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        val executor = createTestExecutor(logger = logger)
+
+        executor.execute(
```
Comment (sergioboi) [inline subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt:52]: And the same about timeout

**Review by sergioboi — CHANGES_REQUESTED** [review]

File: subprojects/gradle/instrumentation-tests-test-fixtures/src/main/kotlin/com/avito/instrumentation/report/listener/FakeTestReporter.kt (hunk 1 above)
Comment (dsvoronin) [reply subprojects/gradle/instrumentation-tests-test-fixtures/src/main/kotlin/com/avito/instrumentation/report/listener/FakeTestReporter.kt:7]: Empty and Stub are the same thing )

Comment (dsvoronin) [reply subprojects/gradle/instrumentation-tests-test-fixtures/src/main/kotlin/com/avito/instrumentation/report/listener/FakeTestReporter.kt:7]: https://stackoverflow.com/a/346440/981330

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/executing/TestExecutor.kt (hunk 3 above)
Comment (dsvoronin) [reply subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/executing/TestExecutor.kt]: removed it completely, not necessary message

File: subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt (hunk 2 above)
Comment (dsvoronin) [reply subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt:44]: thanks, that is a great advice; added

File: subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt (hunk 4 above)
Comment (dsvoronin) [reply subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt]: yep, i changed test names to be more clear

File: subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt (hunk 5 above)
Comment (dsvoronin) [reply subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt:52]: added a timeout for this test too

File: subprojects/gradle/instrumentation-tests-test-fixtures/src/main/kotlin/com/avito/instrumentation/report/listener/FakeTestReporter.kt (hunk 1 above)
Comment (dsvoronin) [reply subprojects/gradle/instrumentation-tests-test-fixtures/src/main/kotlin/com/avito/instrumentation/report/listener/FakeTestReporter.kt:7]: https://github.com/avito-tech/avito-android/discussions/679
Created discussion

File: subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+import org.junit.jupiter.api.Timeout
+import org.junit.jupiter.api.io.TempDir
+import java.io.File
+import java.util.concurrent.TimeUnit.SECONDS
+
+internal class TestExecutorIntegrationTest {
+
+    private val logger = FakeCILogger()
+
+    @Timeout(10, unit = SECONDS)
```
Comment (eugene-krivobokov) [inline subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt:31]: We can define also a default timeout for all tests as a fallback. The same problem can occur in other tests.

`junit.jupiter.execution.timeout.test.method.default`
https://junit.org/junit5/docs/5.5.0-RC2/api/org/junit/jupiter/api/Timeout.html

Comment (dsvoronin) [reply subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt:31]: i've added 5 min global timeout for all tests and lifecycle methods via system property;
tested it with 1 sec timeout, test fails with InterruptedException like on screenshot
maximum test duration on last executions was 4min 11sec

image

Comment (eugene-krivobokov) [reply subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt:31]: If we don't have any other `junit.jupiter.execution.*` properties, I have no idea.

Comment (dsvoronin) [reply subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/executing/TestExecutorIntegrationTest.kt:31]: image

just now, in https://github.com/avito-tech/avito-android/pull/677/commits/11727a5182bcf579221e670d3a2383ff763ee707
sorry for confusion

**Review by sergioboi — APPROVED** [review]

**Pull request merged**